sound/soc/sdca/sdca_interrupts.c

```diff
@@ -0,0 +1,467 @@
+
+		if (!source_mask_byte)
+			continue;
+
+		ret = sdca_interrupt_register_mask_rmw(slave, reg_index,
+						       source_mask_byte,
+						       enable);
+		if (ret < 0)
+			goto unlock;
+	}
+
+	/*
+	 * Almost done, keep track of the combined interrupt mask used to
+	 * filter interrupts in the handler.
+	 */
+	if (enable)
+		interrupt_info->enabled_interrupt_mask |= source_mask;
+	else
+		interrupt_info->enabled_interrupt_mask &= ~source_mask;
+
+unlock:
+	mutex_unlock(&interrupt_info->irqs_lock);
+
+	return ret;
+}
+EXPORT_SYMBOL_NS(sdca_interrupt_enable, SND_SOC_SDCA_IRQ_HANDLER);
+
+void sdca_interrupt_clear_history(struct sdw_slave *slave, u32 preserve_mask)
+{
+	struct sdca_interrupt_info *interrupt_info;
+
+	interrupt_info = slave->sdca_data.interrupt_info;
+
+	/*
+	 * Clear all history except for the interrupts set in preserve_mask.
+	 * This is very useful for SDCA UMP processing, where the
+	 * interrupt is only thrown once when the ownership changes to
+	 * HOST. If the processing happens in a work queue, and a new interrupt
+	 * cancels the work queue, the interrupt will not be signaled again
+	 */
+	interrupt_info->detected_interrupt_mask &= preserve_mask;
+}
+EXPORT_SYMBOL_NS(sdca_interrupt_clear_history, SND_SOC_SDCA_IRQ_HANDLER);
+
+/* helper called with the 'irq_lock' mutex held */
+static int sdca_interrupt_register_handler(struct sdw_slave *slave,
+					   struct sdca_interrupt_info *interrupt_info,
+					   int reg_index)
+{
+	struct device *dev = &slave->dev;
+	unsigned long status2;
+	unsigned long status;
+	const int retry = 3;
+	int count = 0;
+	int bit;
+	int ret;
+
+	ret = sdw_read_no_pm(slave, SDW_SCP_SDCA_INT1 + reg_index);
+	if (ret < 0) {
+		dev_err_ratelimited(dev,
+				    "%s: read of SDW_SCP_SDCA_INT%d failed: %d\n",
+				    __func__, reg_index + 1, ret);
+		goto io_error;
+	}
+	status = ret;
+
+	if (!status)
+		return 0;
+
+	do {
+		/*
+		 * Record detected interrupt sources, source-specific actions
+		 * will be taken after all interrupts have been cleared.
+		 */
+
+		for_each_set_bit(bit, &status, 8) {
+			int index = (reg_index << 3) + bit;
+
+			interrupt_info->detected_interrupt_mask |= BIT(index);
+		}
+
+		/* clear the interrupts for this register */
+		ret = sdw_write_no_pm(slave,
+				      SDW_SCP_SDCA_INT1 + reg_index,
+				      status);
+		if (ret < 0) {
+			dev_err_ratelimited(dev,
+					    "%s: write to SDW_SCP_SDCA_INT%d failed: %d\n",
+					    __func__, reg_index + 1, ret);
+			goto io_error;
+		}
+
+		/*
+		 * The SoundWire specification requires an additional read to make sure
+		 * no interrupts are lost
+		 */
+		ret = sdw_read_no_pm(slave, SDW_SCP_SDCA_INT1 + reg_index);
+		if (ret < 0) {
+			dev_err_ratelimited(dev,
+					    "%s: re-read of SDW_SCP_SDCA_INT%d failed: %d\n",
+					    __func__, reg_index + 1, ret);
+			goto io_error;
+		}
+		status2 = ret;
+
+		/* filter to limit loop to interrupts identified in the first status read */
+		status &= status2;
```

Comment:
`count++;` is missed here

Reply:
oops, yes that's not so good. will review all this.

Reply:
done